sw/device/lib/dif/dif_pwrmgr.h

```diff
@@ -0,0 +1,497 @@
+ * Functions
+ */
+
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_warn_unused_result.h"
+#include <stddef.h>
+#include <stdint.h>
+
+// Header Extern Guard (so header can be used from C and C++)
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+/**
+ * Internal state of a power manager.
+ *
+ * Instances of this struct must be initialized by `dif_pwrmgr_init()` before
+ * being passed to other functions in this library. Its fields should be
+ * considered private and are only provided here so that callers can allocate
+ * it.
+ */
+typedef struct dif_pwrmgr {
+  mmio_region_t base_addr;
+  uint8_t num_wakeup_sources;
+  uint8_t num_reset_sources;
+} dif_pwrmgr_t;
+
+/**
+ * Enumeration for enabling/disabling various functionality.
+ */
+typedef enum dif_pwrmgr_toggle {
+  kDifPwrmgrToggleDisable,
+  kDifPwrmgrToggleEnable,
+} dif_pwrmgr_toggle_t;
+
+/**
+ * Configuration for initializing a power manager.
+ */
+typedef struct dif_pwrmgr_config {
+  /**
+   * Base address of the power manager.
+   */
+  mmio_region_t base_addr;
+  /**
+   * Number of wake-up sources.
+   */
+  uint8_t num_wakeup_sources;
+  /**
+   * Number of reset sources.
+   */
+  uint8_t num_reset_sources;
+} dif_pwrmgr_config_t;
+
+/**
+ * Common return codes for the functions in this library.
+ */
+typedef enum dif_pwrmgr_result {
+  /**
+   * The call succeeded.
+   */
+  kDifPwrmgrOK = 0,
+  /**
+   * A non-specific error occurred and the hardware is in an invalid or
+   * irrecoverable state.
+   */
+  kDifPwrmgrError = 1,
+  /**
+   * The caller supplied invalid arguments but the call did not cause any
+   * side-effects and the hardware is in a valid and recoverable state.
+   */
+  kDifPwrmgrBadArg = 2,
+} dif_pwrmgr_result_t;
+
+/**
+ * Initialize a power manager.
+ *
+ * A power manager must first be initialized by this function before calling
+ * other functions in this library.
+ *
+ * @param config Configuration for initializing a power manager.
+ * @param pwrmgr Internal state of the initialized power manager.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_pwrmgr_result_t dif_pwrmgr_init(dif_pwrmgr_config_t *config,
+                                    dif_pwrmgr_t *pwrmgr);
+
+/**
+ * Enable or disable low power state.
+ *
+ * When enabled, the power manager transitions to low power state on the next
+ * wait for interrupt (WFI) instruction. Since the hardware clears the
+ * corresponding bit automatically, this function must be called before each
+ * transition to low power state.
+ *
+ * @param pwrmgr A power manager.
+ * @param new_state Whether the power manager should transition to low power
+ *                  state on the next WFI instruction.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_pwrmgr_result_t dif_pwrmgr_low_power_enabled_set(
+    const dif_pwrmgr_t *pwrmgr, dif_pwrmgr_toggle_t new_state);
```

Comment:
i'm just curious, would it be clearer if you enumerated all the low power states and what each state's values should be?

Reply:
Can you please elaborate? This function enables transition to low power on the next `WFI`, i.e. the _low power hint_ in the spec. `dif_pwrmgr_low_power_config_set` configures what should happen in low power state. These are two separate functions so that clients can use the same low power configuration as long as they want.

Reply:
let me clarify.  All the functional bits in there technically enable us to build combinations of different power modes.
I was just wondering if it made sense to enumerate the valid power modes for a design, say deep sleep and sleep, and in those structures hardcode what the values might be.

Although thinking about it more, I guess this really should be the responsibility of the code that is one layer above the DIF, and not the DIF itself right?

Reply:
> I guess this really should be the responsibility of the code that is one layer above the DIF, and not the DIF itself right?

Thanks for clarifying. I agree, and TBH I thought about the same thing and I think it will depend on the number possible configurations and how flexible we want this DIF to be. IMHO, leaving this to the clients is a reasonable approach but we can always revisit this if we want the DIF to support only a few fixed configurations.